kart/rev_list_objects.py

```diff
@@ -28,51 +34,99 @@ def _rev_list_objects_command(repo):
 DS_PATH_PATTERN = r"(.+)/\.(sno|table)-dataset/"
 
 
-def rev_list_object_oids(repo, start_commits, stop_commits):
+def rev_list_object_oids(repo, start_commits, stop_commits, pathspecs):
     """
-    Yield all the objects referenced between the start and stop commits as tuples (commit_id, path, object_id).
+    Yield all the objects referenced between the start and stop commits as tuples (commit_id, path, object_id),
+    and which match at least one of the given pathspecs.
     Each object will only be yielded once, so not necessarily at all paths and commits where it can be found.
     """
+    if not pathspecs:
+        return
+
     cmd = [
         *_rev_list_objects_command(repo),
         *start_commits,
         "--not",
         *stop_commits,
+        "--stdin",
     ]
     try:
         p = subprocess.Popen(
             cmd,
+            stdin=subprocess.PIPE,
             stdout=subprocess.PIPE,
             encoding="utf8",
             env=tool_environment(),
         )
+        p.stdin.write(os.linesep.join(["--", *pathspecs]))
+        p.stdin.close()
         yield from _parse_revlist_output(repo, p.stdout)
     except subprocess.CalledProcessError as e:
         raise SubprocessError(
             f"There was a problem with git rev-list: {e}", called_process_error=e
         )
 
 
-def rev_list_blobs(repo, start_commits, stop_commits):
+def get_dataset_pathspecs(repo, start_commits, stop_commits, dirname_filter):
+    """
+    Get the list of dataset paths we need to search in for the objects we are insterested in.
+    Without this, all datasets are searched - even irrelevant ones - which could waste a lot of time.
+    """
+    # TODO - if git rev-list --objects pathspec filtering is fixed, there would be other approaches
+    # we could take, some potentially using only a single call to rev-list.
+    # See https://public-inbox.org/git/CAPJmHpWWJ4sssfG2oym7K=MsT3+KTHQP-QK88nfXNOtfcv07ew@mail.gmail.com/
+    cmd = [
+        *_rev_list_commits_command(repo),
+        *start_commits,
+        "--not",
+        *stop_commits,
+    ]
+    result = set()
+    try:
+        p = subprocess.Popen(
+            cmd,
+            stdout=subprocess.PIPE,
+            encoding="utf8",
+            env=tool_environment(),
+        )
+        for line in p.stdout:
+            commit = repo[line.strip()]
+            for tree_path, tree in all_trees_with_paths_in_tree(commit.tree):
+                if tree_path in result:
```

Comment:
this looks incorrect - shouldn't this be skipping based on tree.oid rather than tree path? Otherwise you'll only iterate a dataset's tree in the first commit you come across

Reply:
This is correct although it may not be doing exactly what you think:
1. all_trees_with_paths_in_tree() iterates over all the trees that aren't "hidden" ie a, a/1, a/2, b, b/1, b/2
2. This code here checks each of those trees to see if they contain any hidden directories that mean they are datasets we should be checking. If it does, we add the path to the result set. But there's no point adding the path to the result set if it's already in the result set - so in that case, we can skip the check to see if it's a dataset, we already know it sometimes is.

I'll add a comment